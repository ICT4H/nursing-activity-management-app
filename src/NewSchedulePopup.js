import React from "react";
import PersonDetails from './PersonDetails'
import SelectOptions from "./SelectOptions";

class NewSchedulePopup extends React.Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.hidePopup = this.hidePopup.bind(this);
    this.allOptions = ["Tablet", "Capsule", "ml", "mg", "Tablespoon"];
  }

  hidePopup() {
    let popup = document.querySelector('.popup');
    popup.style.display = 'none';
    this.props.resetInputFields();
    return true;
  }

  handleChange() {
    let medicine = {
      medicineName: this.refs.medicineName.value,
      dose: this.refs.dose.value,
      unit: this.refs.unit.value
    };
    this.props.onChange(medicine);
  }

  render() {
    let medicine = this.props.medicine;
    return (
        <div className={"popup"}>
          <div className={"popupContent"}>
            <p>Add New Medicine Schedule</p>
            <PersonDetails person={this.props.patient} className={"patientDetails"}/>
            <input
                type={"text"} placeholder={"medicineName"} ref="medicineName"
                value={medicine.medicineName} onChange={this.handleChange}/>

            <input
                type={"text"} placeholder={"dose"} ref="dose"
                value={medicine.dose} onChange={this.handleChange}/>

            <SelectOptions selectedValue={medicine.unit} ref={"unit"} options={this.allOptions}
                           onChange={this.handleChange}/>

            <SaveCancelButtons saveFn={this.hidePopup} cancelFn={this.hidePopup}/>
          </div>
        </div>
    )
  }
}


function SaveCancelButtons(props) {
  return (
      <div>
        <button onClick={props.saveFn} className={"button"} value={"Save"}>Save</button>
        <button onClick={props.cancelFn} className={"button"} value={"Cancel"}>Cancel</button>
      </div>)
}

export default NewSchedulePopup;