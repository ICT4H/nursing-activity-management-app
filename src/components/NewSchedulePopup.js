import React from "react";
import PersonDetails from './PersonDetails'
import SelectOptions from "./SelectOptions";
import SaveCancelButtons from "./SaveCancelButtons";
import {BID, CAP, MG, ML, QD, QID, QOD, TAB, TBSP, TID, TSP} from "../constants";
import {getResultantObject, mapFrequencyToNumber} from "../utils/utility";
import AdministrateTimes from "./AdministrateTimes";
import MedicationInput from "./MedicationInput";
import DosingInstructions from "./DosingInstructions";
import DosingPeriod from "./DosingPeriod";

class NewSchedulePopup extends React.Component {
  constructor(props) {
    super(props);
    this.handleMedicineNameChange = this.handleMedicineNameChange.bind(this);
    this.handleDoseChange = this.handleDoseChange.bind(this);
    this.handleFrequencyChange = this.handleFrequencyChange.bind(this);
    this.handleMedicineUnitChange = this.handleMedicineUnitChange.bind(this);
    this.handleStartingDateChange = this.handleStartingDateChange.bind(this);
    this.handleEndingDateChange = this.handleEndingDateChange.bind(this);
    //todo: Might come from server using an API
    this.medicineUnits = [TAB, CAP, ML, MG, TSP, TBSP];
    this.frequencies = [QD, BID, TID, QID, QOD];
  }

  render() {
    let drug = this.props.drug;
    return (
        <div className={this.getClassName()}>
          <div className="popupContent">
            <p>Add New Medicine Schedule</p>
            <PersonDetails person={this.props.patient} className={"patientDetails"}/>
            <MedicationInput drugName={drug.drugName} onChange={this.handleMedicineNameChange}/>
            <DosingInstructions medicineUnits={this.medicineUnits}
                                handleMedicineUnitChange={this.handleMedicineUnitChange}
                                doseUnit={drug.unit} doseValue={drug.dose}
                                handleDoseChange={this.handleDoseChange}
            />
            <SelectOptions selectedValue={drug.frequency} options={this.frequencies} className="chooseFrequency"
                           ref="frequency" chooseMsg="CHOOSE FREQUENCY" onChange={this.handleFrequencyChange}/>
            <DosingPeriod handleStartingDateChange={this.handleStartingDateChange}
                          handleEndingDateChange={this.handleEndingDateChange}
                          startingDate={drug.startingDate} endingDate={drug.endingDate}
            />
            <AdministrateTimes noOfTimeInputs={mapFrequencyToNumber(drug.frequency)}/>
            <SaveCancelButtons saveFn={this.props.saveFn} cancelFn={this.props.cancelFn}/>
          </div>
        </div>
    )
  }

  handleMedicineNameChange(event) {
    let changedDrugName = {drugName: event.target.value};
    let resultantDrug = getResultantObject(this.props.drug, changedDrugName);
    this.props.onChange(resultantDrug);
  }

  handleDoseChange(event) {
    let changedDose = {dose: event.target.value};
    let resultantDrug = getResultantObject(this.props.drug, changedDose);
    this.props.onChange(resultantDrug);
  }

  handleFrequencyChange(event) {
    let changedFrequency = {frequency: event.target.value};
    let resultantDrug = getResultantObject(this.props.drug, changedFrequency);
    this.props.onChange(resultantDrug);
  }

  handleStartingDateChange(event) {
    let startingDate = new Date(event.target.value);
    let changedStartingDate = {startingDate: startingDate};
    let resultantDrug = getResultantObject(this.props.drug, changedStartingDate);
    this.props.onChange(resultantDrug);
  }

  handleEndingDateChange(event) {
    let startingDate = new Date(event.target.value);
    let changedEndingDate = {endingDate: startingDate};
    let resultantDrug = getResultantObject(this.props.drug, changedEndingDate);
    this.props.onChange(resultantDrug);
  }

  getClassName() {
    return (this.props.hidden) ? "hidden" : "newSchedulePopup";
  }

  handleMedicineUnitChange(event) {
    let changedMedicineUnit = {unit: event.target.value};
    let resultantDrug = getResultantObject(this.props.drug, changedMedicineUnit);
    this.props.onChange(resultantDrug);
  }
}

export default NewSchedulePopup;
