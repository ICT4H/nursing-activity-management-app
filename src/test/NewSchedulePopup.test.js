import React from 'react';
import NewSchedulePopup from '../NewSchedulePopup';
import renderer from 'react-test-renderer';
import moment from "moment/moment";

test('Should have input fields medicineName, dose, unit with given values and save cancel buttons', () => {
  let medicineToPopup = {
    medicineName:"Paracetmol",
    dose:2,
    unit:"Tablet"
  };
  let patient = {
    name: "Cally Cardenas",
    gender:'male',
    age:33
  };

  let component=renderer.create(<NewSchedulePopup medicine={medicineToPopup}
                                                  patient={patient} onChange={()=>true}
                                                  resetInputFields={()=>true}/>);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});