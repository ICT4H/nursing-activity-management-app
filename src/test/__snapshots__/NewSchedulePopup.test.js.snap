// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should have input fields medicineName, dose, unit with given values and save cancel buttons 1`] = `
<div
  className="popup"
>
  <div
    className="popupContent"
  >
    <p>
      Add New Medicine Schedule
    </p>
    <div
      className="patientDetails"
    >
      <p>
        Cally Cardenas
      </p>
      <p>
        male
      </p>
      <p>
        33 yrs
      </p>
    </div>
    <input
      onChange={[Function]}
      placeholder="medicineName"
      type="text"
      value="Paracetmol"
    />
    <input
      onChange={[Function]}
      placeholder="dose"
      type="text"
      value={2}
    />
    <select>
      <option
        value=""
      >
        selectUnit
      </option>
      <option
        value="Tablet"
      >
        Tablet
      </option>
      <option
        value="Capsule"
      >
        Capsule
      </option>
      <option
        value="ml"
      >
        ml
      </option>
      <option
        value="mg"
      >
        mg
      </option>
      <option
        value="Tablespoon"
      >
        Tablespoon
      </option>
    </select>
    <div>
      <button
        className="button"
        onClick={[Function]}
        value="Save"
      >
        Save
      </button>
      <button
        className="button"
        onClick={[Function]}
        value="Cancel"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
